src/python/pants/util/collections.py

```diff
@@ -71,3 +74,49 @@ def ensure_str_list(val: str | Iterable[str], *, allow_single_str: bool = False)
     If `allow_single_str` is True, a single `str` will be wrapped into a `List[str]`.
     """
     return ensure_list(val, expected_type=str, allow_single_scalar=allow_single_str)
+
+
+def partition_sequentially(
+    items: Sequence[_T],
+    *,
+    key: Callable[[_T], str],
+    size_min: int,
+    size_max: int | None = None,
+) -> Iterator[list[_T]]:
+    """Stably partitions the given items into batches of at least size_min.
+
+    The "stability" property refers to avoiding adjusting all batches when a single item is added,
+    which could happen if the items were trivially windowed using `itertools.islice` and an
+    item was added near the front of the list.
+
+    Batches will be capped to `size_max`, which defaults `size_min*2`.
+    """
+
+    # To stably partition the arguments into ranges of at least `size_min`, we sort them, and
+    # create a new batch sequentially once we have the minimum number of entries, _and_ we encounter
+    # an item hash prefixed with a threshold of zeros.
+    zero_prefix_threshold = math.log(size_min // 8, 2)
+    size_max = size_min * 2 if size_max is None else size_max
+
+    batch: list[_T] = []
+
+    def emit_batch() -> list[_T]:
+        assert batch
+        result = list(batch)
+        batch.clear()
+        return result
+
+    keyed_items = []
+    for item in items:
+        keyed_items.append((key(item), item))
+    keyed_items.sort()
+
+    for item_key, item in keyed_items:
+        batch.append(item)
+        if (
+            len(batch) >= size_min
+            and native_engine.hash_prefix_zero_bits(item_key) >= zero_prefix_threshold
```

Comment:
I don't know what the overhead of calling through to Rust is, but are we sure that doing it a lot in a tight loop like this is worth it?

Reply:
It should be very, very low... similar to calling into any builtin Python function. The rust function takes a reference to the Python string, so it shouldn't even be copied at the boundary.

Reply:
```python
>>> timeit.timeit(lambda: native_engine.hash_prefix_zero_bits("example"), number=1000000)
0.19578884500000981
>>> timeit.timeit(lambda: hash("example"), number=1000000)
0.14594443799998658
```